File: 8. Greatest Element on the Right Side/Suffix Max/replaceElements.ts
const replaceElements = (arr: number[]): number[] => {
  let rightMax = -1;
  const ans = new Array(arr.length);
  for (let i = arr.length - 1; i >= 0; i--) {
    const curr = arr[i];
    ans[i] = rightMax;
    rightMax = Math.max(curr, rightMax);
  }
  return ans;
};

// Test cases
console.log(replaceElements([17, 18, 5, 4, 6, 1])); // [18, 6, 6, 6, 1, -1]

File: 8. Greatest Element on the Right Side/Suffix Max/replaceElements.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
  public:
    vector<int> replaceElements(vector<int>& arr) {
        int n = arr.size();
        vector<int> ans(n);
        int rightMax = -1;

        for (int i = n - 1; i >= 0; --i) {
          ans[i] = rightMax;
          rightMax = max(rightMax, arr[i]);
        }
        return ans;
    }
};

int main() {
  Solution solution;
  vector<int> arr = {17, 18, 5, 4, 6, 1};
  vector<int> result = solution.replaceElements(arr);
  for (int i : result) {
    cout << i << " ";
  }
  cout << endl;
  return 0;
}
